File: source/ts/src/core/hora.ts
import { buildEqualSlots, VARA_CHALDEAN_START } from '../utils/slots';
import type { HoraSlot, Unlocalized, UnlocalizedInfo } from '../types/elements';

/** Planet indices are Chaldean order: Sun(0), Venus(1), Mercury(2), Moon(3), Saturn(4), Jupiter(5), Mars(6). */
const DAY_FIRST_HORA = VARA_CHALDEAN_START;

function buildHoras(
  reference: Date,
  durationMs: number,
  firstPlanetIndex: number,
  nameFn: (planetIndex: number) => string,
  count: number,
): Unlocalized<HoraSlot>[] {
  return buildEqualSlots(reference, durationMs, count, (i, start, end) => {
    const planetIndex = (firstPlanetIndex + i) % 7;
    return { start, end, planetIndex, planet: nameFn(planetIndex) };
  });
}

export function computeHora(
  sunrise: Date,
  sunset: Date,
  nextSunrise: Date,
  varaIndex: number,
  nameFn: (planetIndex: number) => string,
): UnlocalizedInfo<HoraSlot> {
  const dayMs   = sunset.getTime()      - sunrise.getTime();
  const nightMs = nextSunrise.getTime() - sunset.getTime();
  const firstDay   = DAY_FIRST_HORA[varaIndex]!;
  const firstNight = (firstDay + 12) % 7;

  return {
    day:   buildHoras(sunrise, dayMs,   firstDay,   nameFn, 12),
    night: buildHoras(sunset,  nightMs, firstNight, nameFn, 12),
  };
}

File: source/go/internal/core/hora.go
package core

import (
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

var horaDayFirst = utils.VaraChaldeanStart

func buildHoras(
	referenceMs int64,
	durationMs float64,
	firstPlanetIndex int,
	nameFn func(planetIndex int) string,
	count int,
) []types.UnlocalizedHoraSlot {
	return utils.BuildEqualSlots(referenceMs, durationMs, count,
		func(i int, startMs, endMs int64) types.UnlocalizedHoraSlot {
			planetIndex := (firstPlanetIndex + i) % 7
			return types.UnlocalizedHoraSlot{
				StartMs: startMs, EndMs: endMs,
				PlanetIndex: planetIndex, Planet: nameFn(planetIndex),
			}
		})
}

func ComputeHora(
	sunriseMs, sunsetMs, nextSunriseMs int64,
	varaIndex int,
	nameFn func(planetIndex int) string,
) types.UnlocalizedHoraInfo {
	dayMs := float64(sunsetMs - sunriseMs)
	nightMs := float64(nextSunriseMs - sunsetMs)
	firstDay := horaDayFirst[varaIndex]
	firstNight := (firstDay + 12) % 7

	return types.UnlocalizedHoraInfo{
		Day:   buildHoras(sunriseMs, dayMs, firstDay, nameFn, 12),
		Night: buildHoras(sunsetMs, nightMs, firstNight, nameFn, 12),
	}
}
